internals/secrethub/run.go

```diff
@@ -339,12 +340,18 @@ type envvarTpls struct {
 	lineNo int
 }
 
+type secretReaderNotAllowed struct{}
+
+func (sr secretReaderNotAllowed) ReadSecret(path string) (string, error) {
+	return "", ErrSecretsNotAllowedInKey
+}
+
 // Env injects the given secrets in the environment values and returns
 // a map of the resulting environment.
 func (t envTemplate) Env(secrets map[string]string, sr tpl.SecretReader) (map[string]string, error) {
 	result := make(map[string]string)
 	for _, tpls := range t.envVars {
-		key, err := tpls.key.Evaluate(t.templateVars, sr)
+		key, err := tpls.key.Evaluate(t.templateVars, secretReaderNotAllowed{})
```

Comment:
Seeing that we're not passing a pointer here, led me to looking at other usages of `Evaluate()`. Seems that we're always passing a value and in case of a normal `secretReader` that value contains the value of a `secrethub.Client`. Meaning that we're copying the client. Is that something we want?

Reply:
Ahh, no, it's not. We want to have a pointer to a client, as the client can cache etc.

Actually, we probably want the SecretReader to accept a function that creates a client and have it call that function on the first call to ReadSecret. By doing so, we postpone creating the client, which postpones prompting the user for their passphrase.
For example, in case of a parse error, we don't want to prompt the user for their passphrase.
And for templates without any secret, we then won't prompt at all :+1:.

I'll change this in a new PR.